Спринт 8

PrettyDG opened the pull request:
Появился новый метод getPrioritizedTasks - он выводит задачи в заданном порядке. Появилась проверка пересечения задач при добавлении. Если задача по времени пересекается с другой - она не будет добавлена. Заменены циклы 'for' на Stream API.
Добавилось множество различных тестов.

PrettyDG committed:
Добавление времени старта и выполнения для задач.
Появился новый метод getPrioritizedTasks - он выводит задачи в заданном порядке.
Появилась проверка пересечения задач при добавлении. Если задача по времени пересекается с другой - она не будет добавлена.
Заменены циклы 'for' на Stream API.
Добавилось множество различных тестов.

PrettyDG committed:
Визуальные изменения1

PrettyDG committed:
Визуальные изменения2

PrettyDG commented:
Для корректного прохождения всех тестов пришлось добавить метод 'clearEverything' в InMemoryTaskManager - он очищает мапы задач и приравнивает к нулю айди для задач. Можно ли его использовать? Или необходимо сделать так, чтобы и без него тесты корректно проходили?

avfyodorov commented:
> Для корректного прохождения всех тестов пришлось добавить метод 'clearEverything' в InMemoryTaskManager - он очищает мапы задач и приравнивает к нулю айди для задач. Можно ли его использовать? Или необходимо сделать так, чтобы и без него тесты корректно проходили?

Лучше было бы в тестах, в метод BeforeEach, Добавить создание соответствующего менеджера. Пусть тебе создаётся каждый раз. 

	@BeforeEach
	public void setUp() {
		taskManager = new InMemoryTaskManager();
		initTasks();
	}

avfyodorov commented on src/models/Epic.java:
```diff
@@ -1,20 +1,59 @@
+                }
+            }
+            setStartTime(minTime);
+            setDuration(Duration.between(minTime, maxTime));
```
setDuration(Duration.between(minTime, maxTime));

Не совсем так. Для Эпика нужно суммировать продолжительность всех его подзадач. По техническому заданию: 

.==============
А вот с классом Epic нужно поработать дополнительно. Продолжительность эпика — сумма продолжительностей всех его подзадач.

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -42,11 +50,31 @@ public static FileBackedTaskManager loadFromFile(File file) {
         return manager;
     }
 
+    public TreeSet<Task> getPrioritizedTasks() {
```
Здесь нужно список возвращать, а выглядеть метод должен приблизительно так: 


	@Override
	public List<Task> getPrioritizedTasks() {
		return new ArrayList<>(prioritizedTasks);
	}

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -1,17 +1,20 @@
     private final HashMap<Integer, Subtask> subtaskHash = new HashMap<>();
-    private static int idNumberForTasks = 0;
     private final HistoryManager history = Managers.getDefaultHistory();
 
```
Правильнее было бы завести здесь список задач, отсортированых по времени и динамически добавлять сюда задачи и подзадачи. И, соответственно, удалять из списка при удалении. 

	protected final Set<Task> prioritizedTasks = new TreeSet<>(Comparator......................

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -23,6 +26,13 @@ public ArrayList<Task> getAllTasks() {
         return allTasks;
     }
 
+    public void clearEverything() {
```
Лучше Перед каждым тестом создавать свой менеджер задач и набор тестовых задач

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -48,9 +58,8 @@ public DefaultTask deleteDefaultTaskByID(int id) throws ManagerSaveException {
-            history.remove(id);
-        }
+        defaultTasksHash.keySet().stream()
+                .forEach(history::remove);
```
А Нужен ли здесь STREAM?

Почему нельзя сделать так: 

defaultTasksHash.keySet()
                .forEach(history::remove);

avfyodorov commented on src/controllers/InMemoryTaskManager.java:
```diff
@@ -48,9 +58,8 @@ public DefaultTask deleteDefaultTaskByID(int id) throws ManagerSaveException {
 
     @Override
     public void deleteAllDefaultTasks() throws ManagerSaveException {
```
throws ManagerSaveException 

Здесь и дальше. Можно обойтись без этого ключевого слова, поскольку исключение не проверяемое, оно само пробросится. 

(ManagerSaveException extends RuntimeException)

avfyodorov reviewed (commented):
Добрый день, Данила! 
В целом работа хорошая, единственное, нужно будет доделать хранение и работу с отсортированным по времени списком задач.

PrettyDG committed:
Переделан расчёт duration и endTime для EpicTask, также for() заменена на stream().
getPrioritizedTasks() теперь возвращает List<Task>, задачи попадают в TreeSet prioritizedTasks сразу после создания и пропадают после удаления задачи.
Удалён метод clearEverything(), взамен используется @BeforeEach.

PrettyDG commented:
Благодарю за подсказку с 'BeforeEach'! Совсем забыл о его существовании. 
Удалось реализовать правильное прохождение всех тестов без использования вспомогательных методов для очистки предыдущих результатов.

avfyodorov reviewed (commented):
Добрый день, Данила!

Замечаний нет.
Работа принята.
